text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,749 @@
+enum BlockHeaderIter<'a> {
+    Parent(&'a Block),
+    RootState(&'a Block),
+    Deploys(&'a Block, usize),
+    Done,
+}
+
+impl<'a> OutIterable for &'a BlockHeader {
+    type OutIter = BlockHeaderIter<'a>;
+
+    fn iter_outgoing(self) -> Self::OutIter {
+        BlockHeaderIter::Parent(self)
+    }
+}
+
+impl<'a> Iterator for BlockHeaderIter<'a> {
+    type Item = ObjectHash;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut rv = None;
+
+        *self = match *self {
+            BlockHeaderIter::Parent(block) => {
+                rv = block.parent_hash.clone();
+                BlockHeaderIter::RootState(block)
+            }
+            BlockHeaderIter::RootState(block) => {
+                rv = block.root_state.clone();
+                BlockHeaderIter::Deploys(block, 0)
+            }
+            BlockHeaderIter::Deploys(block, idx) if idx < block.deploys.len() => {
+                rv = block.deploys[idx].clone();
+                BlockHeaderIter::Deploys(block, idx + 1)
+            }
+            BlockHeaderIter::Deploys(block, idx)
+            | BlockHeaderIter::Done => BlockHeaderIter::Done,
+        };
+
+        rv
+    }
+}
+```
+
+_NOTE: The above logic is necessary to store and send arbitrary objects using this proposed model; the following logic describes optional convenience methods implemented on `Object` to improve the ergonomics for component developers._
+
+```rust
+impl Object {
+    fn as_block(&self) -> Option<&Block> {
+        match self {
+            Object::Block(ref block) => Some(block),
+            _ => None
+        }
+    }
+
+    // ...
+}
+```
+
+### Storage implementation
+
+Using this proposed model, the storage component decomplects into a single `ObjectStore`, with a simplified set of effects.
+
+```rust
+/// Retrieve multiple objects.
+// This method always uses the heap, since a single is already quite large.
+// Does not return any dependencies, may return less objects than requested.
+pub async fn get_objects(self, object_hash: Vec<ObjectHash>) -> HashMap<ObjectHash, ImmutableObject>;
+
+/// Retrieve multiple objects, including their dependencies.
+///
+/// If specified, stops at a specific depth. Returns `None` if any requested
+/// object was incomplete until `max_depth`.
+pub async fn get_objects_with_dependencies(self, object_hash: Vec<ObjectHash>, max_depth: Option<NonZeroUsize>) -> Option<HashMap<ObjectHash, ImmutableObject>>>;
+
+/// Store multiple objects.
+///
+/// Objects are stored in the reverse order given, so dependencies should be
+/// listed behind the actual object.
+pub async fn put_objects(self, objects: Vec<ImmutableObject>);
+```
+
+The `ObjectStore` continues to use LMDB as a backend, but calculates keys upfront (we might also consider writing to a temporary key and renaming it) and stores the object under a precalculated key. For convenience, given a proper `Hash` trait implementation for `ObjectHash`, a hashmap might be returned.
```

Comment:
> Aren't these just the objects' hashes?

Yes, that's what I meant =).

> we might also consider writing to a temporary key and renaming it

If we were to store an object that is brand new and has never had its hash created, it would have to be serialized to do so. So instead of creating another copy in memory, we write it to a temporary key, calculating the hash as we write it. Once that's done, we rename the temporary key to match the hash. All in all it's purely an optimization that saves copying an object on some occasions.